src/pcp/types.rs

```diff
@@ -108,25 +107,38 @@ impl<F: FieldElement> Type for Count<F> {
 ///
 /// [BBCG+19]: https://ia.cr/2019/188
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub struct Sum<F> {
+pub struct Sum<F: FieldElement> {
     bits: usize,
+    one: F::Integer,
+    max_summand: F::Integer,
     range_checker: Vec<F>,
 }
 
 impl<F: FieldElement> Sum<F> {
     /// Return a new [`Sum`] type parameter. Each value of this type is an integer in range `[0,
     /// 2^bits)`.
     pub fn new(bits: usize) -> Result<Self, PcpError> {
-        let max_bits = size_of::<F::Integer>() << 3;
-        if bits > max_bits {
+        let bits_int = F::Integer::try_from(bits).map_err(|err| {
+            PcpError::Encode(format!(
+                "bit length ({}) cannot be represented as a field element: {:?}",
+                bits, err,
+            ))
+        })?;
+
+        if F::modulus() >> bits_int == F::Integer::try_from(0).unwrap() {
             return Err(PcpError::Encode(format!(
-                "bits ({}) exceeds bit length of the field's integer representation ({})",
-                bits, max_bits,
+                "bit length ({}) exceeds field modulus",
+                bits,
             )));
         }
 
+        let one = F::Integer::try_from(1).unwrap();
```

Comment:
I feel like we do this in a lot of places. Maybe we could clean up a bunch of unwraps and avoid caching the value of 1 in struct `Sum` if we added a bound to `FieldElement::Integer` that requires some trait that exposes infallible `one()` and `zero()` methods. I wouldn't do that in this PR, though.

Reply:
Ugh, I always forget about `F::Integer::from(F::one())` and `F::Integer::from(F::zero())`. I did this here, but will wait to fix it elsewhere for a future PR.